Intelligent Core tracing (#22633)
newrelic/docs-website: src/content/docs/distributed-tracing/core-tracing/controlling-trace-ingest.mdx

---
title: Control distributed trace ingest with samplers
tags:
  - Understand dependencies
  - Distributed tracing
  - Span configuration
  - Sampling

>   - APM agents

Reviewer: ```suggestion
```

**Common pattern**: Respect upstream decision (don't sample):

```yaml
distributed_tracing:
  sampler:
    remote_parent_not_sampled:
      always_off  # Don't sample traces that upstream rejected
```

**Alternative pattern**: Sample some traces even if upstream rejected them (use cautiously):

```yaml
distributed_tracing:
  sampler:
    remote_parent_not_sampled:
      trace_id_ratio_based:
        ratio: 0.05  # Sample 5% even if upstream said no
```

<Callout variant="caution">
  Overriding upstream "not sampled" decisions can create fragmented traces where only some services are present. Use this pattern only when you have a specific need to see activity in a downstream service even when upstream doesn't think the trace is interesting.
</Callout>

## Sampler types [#sampler-types]

APM agents provide four sampler types, each appropriate for different scenarios.

### Adaptive sampler

**How it works**: Dynamically adjusts sampling to capture approximately a target number of traces per minute, regardless of traffic volume.

**Default behavior**: Most agents default to adaptive sampling with a target of 10 traces per minute.

**When to use**:

> * You want consistent trace volume regardless of traffic spikes or drops

Reviewer: ```suggestion
* You want a consistent number of transactions sampled  regardless of traffic spikes or drops.
```

> * You want automatic adjustment to traffic patterns

Reviewer: ```suggestion
* You want dynamic adjustment to space out sampling based on service throughput.
```

* **Maximum**: Most agents support up to 120 traces per minute

**How it adapts**:
* Low traffic (10 requests/min): Samples 100% of requests to hit target
* Medium traffic (100 requests/min): Samples ~20% to hit target
* High traffic (1000 requests/min): Samples ~2% to hit target

**Example**: Capture 60 traces per minute for complete service map coverage:

```yaml
distributed_tracing:
  sampler:
    root:
      adaptive:
        sampling_target: 60
    remote_parent_sampled:
      always_on
```

### Ratio-based sampler (trace_id_ratio_based)

**How it works**: Samples a fixed percentage of traces based on the trace ID. Because the sampling decision is based on trace ID, the same trace will get the same decision across all services using ratio-based sampling with the same ratio.

**When to use**:
* You need a consistent percentage of traffic sampled
* You want deterministic sampling (same traces always sampled)
* You're coordinating sampling across multiple services

**Configuration**:

```yaml
distributed_tracing:
  sampler:
    root:
      trace_id_ratio_based:
        ratio: 0.1  # 10% of traces
```

**Key parameters**:
* `ratio`: Decimal between 0.0 (0%) and 1.0 (100%)

> **Important characteristic**: The trace ID determines whether a trace is sampled. If multiple services use the same ratio, they'll make consistent decisions for the same trace.

Reviewer: ```suggestion
**Important characteristic**: The trace ID determines whether a transaction is sampled. If multiple services use the same ratio, they'll make consistent decisions for the same distributed trace.
```

> **Example**: Sample 10% of root traces and 100% of upstream-sampled traces:

Reviewer: Let's change it to the following to demo the case where you  actually don't want to sample all remote parent sampled traces (avoid adaptive sampling behavior).
```suggestion
**Example**: Sample 10% of root traces and 50% of upstream-sampled traces:
```

```yaml
distributed_tracing:
  sampler:
    root:
      trace_id_ratio_based:
        ratio: 0.1
    remote_parent_sampled:

>       always_on

Reviewer: ```suggestion
      trace_id_ratio_based:
          ratio: .5
```

```

### Always-on sampler

**How it works**: Samples 100% of traces matching this context.

**When to use**:
* You need complete trace coverage for a specific context
* You're using partial granularity to keep data volume manageable
* You're sampling traces that upstream already selected (common pattern)

**Configuration**:

```yaml
distributed_tracing:
  sampler:
    remote_parent_sampled:
      always_on  # Capture all upstream-sampled traces
```

> **Warning**: Using `always_on` for root context will capture every trace, which can generate significant data volume for high-traffic services. Consider using partial granularity if you enable `always_on`.

Reviewer: I think this is true regardless of whether you configure it for root or remote parent sampled, etc:
```suggestion
**Warning**: Using `always_on` will capture every transaction for the given context, which can generate significant data volume for high-traffic services. Consider capturing partial granularity  instead of full granularity distributed traces for a given section if you configure `always_on`.
```

> **Warning**: Using `always_on` for root context will capture every trace, which can generate significant data volume for high-traffic services. Consider using partial granularity if you enable `always_on`.

Reviewer: I think this needs to be put inside a callout block.

### Always-off sampler

**How it works**: Samples 0% of traces matching this context.

**When to use**:
* You want to respect upstream "not sampled" decisions

> * You're disabling sampling for a specific context

Reviewer: For consistency with always on let's move this bullet point to be first:
```suggestion
* You want to disable sampling for a specific context
```
